src/herder/Upgrades.cpp

```diff
@@ -1029,4 +1242,201 @@ Upgrades::applyReserveUpgrade(AbstractLedgerTxn& ltx, uint32_t newReserve)
+
+    ConfigUpgradeSet upgradeSet;
+    try
+    {
+        xdr::xdr_from_opaque(bytes, upgradeSet);
+    }
+    catch (xdr::xdr_runtime_error&)
+    {
+        return nullptr;
+    }
+
+    return std::shared_ptr<ConfigUpgradeSetFrame>(
+        new ConfigUpgradeSetFrame(upgradeSet, key));
+}
+
+ConfigUpgradeSetFrame::ConfigUpgradeSetFrame(
+    ConfigUpgradeSet const& upgradeSetXDR, ConfigUpgradeSetKey const& key)
+    : mConfigUpgradeSet(upgradeSetXDR)
+    , mKey(key)
+    , mValidXDR(isValidXDR(upgradeSetXDR, key))
+{
+}
+
+bool
+ConfigUpgradeSetFrame::isValidXDR(ConfigUpgradeSet const& upgradeSetXDR,
+                                  ConfigUpgradeSetKey const& key) const
+{
+    if (key.contentHash != sha256(xdr::xdr_to_opaque(upgradeSetXDR)))
+    {
+        CLOG_DEBUG(Herder,
+                   "Got bad configUpgradeSet. Does not match hash in key {}",
+                   hexAbbrev(key.contentHash));
+        return false;
+    }
+
+    if (upgradeSetXDR.updatedEntry.empty())
+    {
+        CLOG_DEBUG(Herder, "Got bad configUpgradeSet {}: no entries updated",
+                   hexAbbrev(key.contentHash));
+        return false;
+    }
+
+    if (!std::is_sorted(
+            upgradeSetXDR.updatedEntry.begin(),
+            upgradeSetXDR.updatedEntry.end(),
+            [](ConfigSettingEntry const& a, ConfigSettingEntry const& b) {
+                return a.configSettingID() < b.configSettingID();
+            }))
+    {
+        CLOG_DEBUG(Herder,
+                   "Got bad configUpgradeSet {}: the entries are not ordered",
+                   hexAbbrev(key.contentHash));
+        return false;
+    }
+    if (std::adjacent_find(
+            upgradeSetXDR.updatedEntry.begin(),
+            upgradeSetXDR.updatedEntry.end(),
+            [](ConfigSettingEntry const& a, ConfigSettingEntry const& b) {
+                return a.configSettingID() == b.configSettingID();
+            }) != upgradeSetXDR.updatedEntry.end())
+    {
+        CLOG_DEBUG(Herder, "Got bad configUpgradeSet {}: duplicate entry",
+                   hexAbbrev(key.contentHash));
+        return false;
+    }
+    return true;
+}
+
+ConfigUpgradeSet const&
+ConfigUpgradeSetFrame::toXDR() const
+{
+    return mConfigUpgradeSet;
+}
+
+ConfigUpgradeSetKey const&
+ConfigUpgradeSetFrame::getKey() const
+{
+    return mKey;
+}
+
+LedgerKey
+ConfigUpgradeSetFrame::getLedgerKey(ConfigUpgradeSetKey const& upgradeKey)
+{
+    SCVal v;
+    v.type(SCV_BYTES);
+    v.bytes().insert(v.bytes().begin(), upgradeKey.contentHash.begin(),
+                     upgradeKey.contentHash.end());
+
+    LedgerKey lk;
+    lk.type(CONTRACT_DATA);
+    lk.contractData().contractID = upgradeKey.contractID;
+    lk.contractData().key = v;
+    return lk;
+}
+
+Hash const&
+ConfigUpgradeSetFrame::getHash() const
+{
+    return mKey.contentHash;
+}
+
+bool
+ConfigUpgradeSetFrame::upgradeNeeded(AbstractLedgerTxn& ltx) const
+{
+    for (auto const& updatedEntry : mConfigUpgradeSet.updatedEntry)
```

Comment:
missing version check (`load` below will blow up otherwise)

Reply:
We can't get here without a valid `ConfigUpgradeSetFrame`, which requires a `loadWithoutRecord`, so `makeFromkey` needs a version check. This leads me to another question - should this call `loadWithoutRecord` instead of `load`?

Reply:
Added a version check